scripts/word_embeddings/tools/cooccur.cc

```diff
@@ -0,0 +1,369 @@
+
+  void push_back(T item) {
+    if (size() < capacity()) {
+      _data.push_back(item);
+    } else {
+      _data.at(_pointer) = item;
+      _pointer = (_pointer + 1) % _capacity;
+    }
+  }
+
+  void clear() {
+    _data.clear();
+    _pointer = 0;
+  }
+
+private:
+  const size_t _capacity;
+  std::vector<T> _data;
+  size_t _pointer = 0;
+
+  // Ranges TS based iterator over circular buffer
+  friend ranges::range_access;
+  struct cursor {
+    cursor() = default;
+    cursor(size_t pointer, size_t counter, const CircularBuffer *parent)
+        : pointer(pointer), counter(counter), parent(parent) {}
+
+    T read() const { return parent->_data.at(pointer); }
+    bool equal(const cursor &other) const { return counter == other.counter; }
+    void next() {
+      // LIFO
+      if (pointer == 0) {
+        pointer = parent->size() - 1;
+      } else {
+        pointer -= 1;
+      }
+      counter++;
+    }
+
+    size_t pointer;
+    size_t counter;
+    const CircularBuffer *parent;
+  };
+  cursor begin_cursor() const {
+    if (size() == 0)
+      return cursor(_pointer, 0, this);
+    else if (_pointer == 0)
+      return cursor(size() - 1, 0, this);
+    else
+      return cursor(_pointer - 1, 0, this);
+  }
+  cursor end_cursor() const { return cursor(_pointer, size(), this); }
+};
+
+// * Input
+auto ReadVocab() {
+  std::string word;
+  std::string count;
+  int rank{0};
+  Vocab vocab;
+  while (std::cin >> word) {
+    std::cin >> count;
+    vocab[word] = {rank, std::stoi(count)};
+    rank++;
+  }
+  return vocab;
+}
+// * Co-occurence matrix construction
+std::mutex paths_m;
+std::mutex matrices_m;
+
+void ReadMatrix(std::queue<fs::path> &paths, queue<Matrix> &matrices,
+                const Vocab &vocab, const Arguments &args, uint32_t seed) {
+  assert(seed > 0);
+  std::string line;
+  CircularBuffer<uint32_t> history(args.window_size);
+  std::unique_ptr<Matrix> m = std::make_unique<Matrix>();
+
+  // Prepare subsampling
+  std::random_device r;
+  std::default_random_engine random_engine(r());
+  std::uniform_real_distribution<float> uniform_dist(0, 1);
+  std::vector<double> idx_to_pdiscard;
+  if (args.subsample) {
+    auto counts = vocab | rsv::values | rsv::values;
+    double sum_counts = std::accumulate(counts.begin(), counts.end(), 0);
+    double t = 1E-4;
+    for (const double count : vocab | rsv::values | rsv::values) {
+      idx_to_pdiscard.push_back(1 - std::sqrt(t / (count / sum_counts)));
+    }
+  }
+
+  while (true) {
+    fs::path path;
+    {
+      std::scoped_lock lock(paths_m);
+      if (paths.empty()) {
+        break;
+      }
+      path = paths.front();
+      paths.pop();
+    }
+
+    std::ifstream in{path};
+    if (!in.is_open()) {
+      throw std::invalid_argument(path.string() + " cannot be opened!");
+    }
+    while (std::getline(in, line)) {
+      history.clear(); // Discard context from other lines
+
+      auto words = line | rsv::split((int (*)(int)) & std::isspace);
```

Comment:
What does the `|` operator do for string?

Reply:
I also don't quite understand what `(int (*)(int)) & std::isspace` means..

Reply:
`|` here is part of [Ranges TS](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1037r0.pdf). It will be part of C++20 according to the Standards Meeting in Rapperswil this summer (included here via `#include "range/v3/all.hpp"` which runs on C++11)

Reply:
`line | rsv::split((int (*)(int)) & std::isspace);` returns a range over all tokens in the line.

> Given a source range and a delimiter specifier, split the source range into a range of ranges using the delimiter specifier to find the boundaries. The delimiter specifier can be a value, a subrange, a predicate, or a function. The predicate should take an single argument of the range's reference type and return true if and only if the element is part of a delimiter. The function should accept current/end iterators into the source range and return make_pair(true, iterator_past_the_delimiter) if the current position is a boundary; otherwise, make_pair(false, cur). The delimiter character(s) are excluded from the resulting range of ranges

Unfortunately, the lack of https://wg21.link/P0919 means we need to copy each range (token) to a `std::string` in the end, but once https://wg21.link/P0919 is supported, using range with string_view would avoid any copies here.

Some more example code for splitting ranges is at https://github.com/ericniebler/range-v3/blob/f662cbc3f8643cc4d8daaab0b53b9b1daddf663b/test/view/split.cpp#L146-L157